core/common/src/main/java/alluxio/util/JvmPauseMonitor.java

```diff
@@ -0,0 +1,207 @@
+      GcTimes diff = gcTimesAfterSleep.get(name).subtract(
+          gcTimesBeforeSleep.get(name));
+      if (diff.mGcCount != 0) {
+        gcDiffs.add("GC pool '" + name + "' had collection(s): "
+            + diff.toString());
+      }
+    }
+
+    String ret = "Detected pause in JVM or host machine (eg GC): "
+        + "pause of approximately " + extraSleepTime + "ms\n";
+    if (gcDiffs.isEmpty()) {
+      ret += "No GCs detected";
+    } else {
+      ret += Joiner.on("\n").join(gcDiffs);
+    }
+    return ret;
+  }
+
+  private Map<String, GcTimes> getGcTimes() {
+    Map<String, GcTimes> map = Maps.newHashMap();
+    List<GarbageCollectorMXBean> gcBeans =
+        ManagementFactory.getGarbageCollectorMXBeans();
+    for (GarbageCollectorMXBean gcBean : gcBeans) {
+      map.put(gcBean.getName(), new GcTimes(gcBean));
+    }
+    return map;
+  }
+
+  private static final class GcTimes {
+    private GcTimes(GarbageCollectorMXBean gcBean) {
+      mGcCount = gcBean.getCollectionCount();
+      mGcTimeMillis = gcBean.getCollectionTime();
+    }
+
+    private GcTimes(long count, long time) {
+      mGcCount = count;
+      mGcTimeMillis = time;
+    }
+
+    private GcTimes subtract(GcTimes other) {
+      return new GcTimes(this.mGcCount - other.mGcCount,
+          mGcTimeMillis - other.mGcTimeMillis);
+    }
+
+    @Override
+    public String toString() {
+      return "count=" + mGcCount + " time=" + mGcTimeMillis + "ms";
+    }
+
+    private long mGcCount;
+    private long mGcTimeMillis;
+  }
+
+  private class Monitor implements Runnable {
+    @Override
+    public void run() {
+      StopWatch sw = new StopWatch();
+      Map<String, GcTimes> gcTimesBeforeSleep = getGcTimes();
+      while (mShouldRun) {
+        sw.reset().start();
+        try {
+          Thread.sleep(mGcSleepINtervalMs);
+        } catch (InterruptedException ie) {
+          return;
+        }
+        long extraSleepTime = sw.now(TimeUnit.MILLISECONDS) - mGcSleepINtervalMs;
+        Map<String, GcTimes> gcTimesAfterSleep = getGcTimes();
+
+        if (extraSleepTime > mWarnThresholdMs) {
+          ++mNumGcWarnThresholdExceeded;
+          LOG.warn(formatMessage(
+              extraSleepTime, gcTimesAfterSleep, gcTimesBeforeSleep));
+        } else if (extraSleepTime > mInfoThresholdMs) {
+          ++mNumGcInfoThresholdExceeded;
+          LOG.info(formatMessage(
+              extraSleepTime, gcTimesAfterSleep, gcTimesBeforeSleep));
+        }
+        mTotalGcExtraSleepTime += extraSleepTime;
+        gcTimesBeforeSleep = gcTimesAfterSleep;
+      }
+    }
+  }
+
+  /**
+   * Simple 'main' to facilitate manual testing of the pause monitor.
+   *
+   * This main function just leaks memory into a list. Running this class
+   * with a 1GB heap will very quickly go into "GC hell" and result in
+   * log messages about the GC pauses.
+   * @param  args arg from command
+   */
+  public static void main(String []args) throws Exception {
```

Comment:
can you move this method to a test?

Reply:
@ifcharming I have commit the test case. But I am not sure whether it is appropriate. Please take a look.

Reply:
Could you remove this method?

Reply:
Sure!